syntax = "proto3";

package recommendations;

service Recommender {
  rpc GetUserRecommendations (UserRequest) returns (RecommendationResponse);
  rpc GetTitleRelevant (TitleRequest) returns (RecommendationResponse);
  rpc TrainModel (TrainRequest) returns (TrainResponse);
  rpc RegisterInteraction (InteractionRequest) returns (InteractionResponse);
}

message UserRequest {
  int32 user_id = 1;
  int32 limit = 2;
  bool filter_viewed = 3;
}
message TitleRequest {
  int32 title_id = 1;
}
message RecommendationResponse {
  repeated int32 item_ids = 1;
}

message TrainRequest {
  bool force_retrain = 1;
}

message TrainResponse {
  bool success = 1;
  string message = 2;
  int32 version = 3;
}

message InteractionRequest {
  int32 user_id = 1;
  int32 item_id = 2;
  string interaction_type = 3;
  float weight = 4;
}

message InteractionResponse {
  bool success = 1;
  string message = 2;
}